include/translated-post.php

```diff
@@ -259,45 +259,76 @@ public function current_user_can_read( $post_id, $context = 'view' ) {
-			'suppress_filters' => 0, // To make the post_fields filter work
-			'lang'             => 0, // Avoid admin language filter
-			'numberposts'      => 20, // Limit to 20 posts
-			'post_status'      => 'any',
-			'post_type'        => $type,
-			'tax_query'        => array(
-				array(
-					'taxonomy' => $this->tax_language,
-					'field'    => 'term_taxonomy_id', // WP 3.5+
-					'terms'    => $lang->term_taxonomy_id,
-				),
-			),
-		);
+		global $wpdb;
 
 		/**
-		 * Filter the query args when auto suggesting untranslated posts in the Languages metabox
-		 * This should help plugins to fix some edge cases
+		 * Filters the limit of queried posts during a search of untranslated ones.
+		 * Formerly `pll_ajax_posts_not_translated_args` when `get_posts()` was used.
 		 *
-		 * @see https://wordpress.org/support/topic/find-translated-post-when-10-is-not-enough
+		 * @since 3.4
 		 *
-		 * @since 1.7
-		 *
-		 * @param array $args WP_Query arguments
+		 * @param int $limit Limit to searched posts, default 20.
 		 */
-		$args  = apply_filters( 'pll_ajax_posts_not_translated_args', $args );
-		$posts = get_posts( $args );
+		$limit = apply_filters( 'pll_ajax_posts_not_translated_limit', 20 );
+
+		$search_like = '%' . $wpdb->esc_like( $search ) . '%';
+
+		$search_query = $wpdb->prepare(
+			"INNER JOIN {$wpdb->term_relationships} AS tr1 ON ({$wpdb->posts}.ID = tr1.object_id)
+			AND tr1.term_taxonomy_id IN (%d)
+			AND (({$wpdb->posts}.post_title LIKE %s)
+				OR ({$wpdb->posts}.post_excerpt LIKE %s)
+				OR ({$wpdb->posts}.post_content LIKE %s))
+			AND {$wpdb->posts}.post_type = %s
+			AND (({$wpdb->posts}.post_status <> 'trash'
+			AND {$wpdb->posts}.post_status <> 'auto-draft'))",
+			$lang->term_taxonomy_id,
+			$search_like,
+			$search_like,
+			$search_like,
+			$type
+		);
+
+		$untranslated_like = '%' . $wpdb->esc_like( $untranslated_in->slug ) . '%';
+
+		$not_translated_query = $wpdb->prepare(
+			"WHERE {$wpdb->posts}.ID NOT IN (
+				SELECT {$wpdb->posts}.ID FROM {$wpdb->posts}
+					LEFT JOIN {$wpdb->term_relationships} AS tr2 ON ({$wpdb->posts}.ID = tr2.object_id)
+					INNER JOIN {$wpdb->term_taxonomy} AS tt ON (tt.term_taxonomy_id = tr2.term_taxonomy_id)
+						AND (tt.taxonomy = 'post_translations')
+						AND tt.description LIKE %s
+			)",
+			$untranslated_like
+		);
 
-		foreach ( $posts as $post ) {
-			if ( $post instanceof WP_Post && ! $this->get_translation( $post->ID, $untranslated_in ) && $this->current_user_can_read( $post->ID, 'edit' ) ) {
-				$return[] = $post;
+		$limit_query = $wpdb->prepare(
+			'LIMIT 0, %d',
+			$limit
+		);
+
+		$sql = "SELECT * FROM {$wpdb->posts}
+				{$search_query}
+                {$not_translated_query}
+                GROUP BY {$wpdb->posts}.ID
+                ORDER BY {$wpdb->posts}.post_date DESC
+                {$limit_query}";
+
+		/**
+		 * Already prepared above.
+		 */
+		$posts = $wpdb->get_results( $sql ); // phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```

Comment:
I wonder if it would be absolutely unreadable to write all in only one operation to avoid this  `phpcs:ignore` as I don't see any special operation which could not gor through `prepare()`.

Reply:
Done in 21f6e6f2, LGTM. What do you think?